---
title: "Mars-2020"
output: html_document
date: "2024-01-14"
---

```{r setup, include=FALSE} 
knitr::opts_chunk$set(echo= TRUE, warning = FALSE, message = FALSE) 
```

***********
# 1. IMPORT
***********

## 1.1. Libraries

```{r library-import}
library(MicrobiotaProcess) # an R package for analysis, visualization and biomarker discovery of Microbiome
library(phyloseq) # Handling and analysis of high-throughput microbiome census data.
library(ggplot2) # Create Elegant Data Visualisations Using the Grammar of Graphics.
library(tidyverse) # Easily Install and Load the 'Tidyverse'.
library(vegan) # Community Ecology Package.
library(coin) # Conditional Inference Procedures in a Permutation Test Framework.
library(reshape2) # Flexibly Reshape Data: A Reboot of the Reshape Package.
library(ggnewscale) # Multiple Fill and Colour Scales in 'ggplot2'
library(patchwork)
library(VennDiagram)
library(UpSetR)
library(ggtree)
library(ggsignif)
library(cowplot)
library(microViz)
```


```{r path-directories, echo=FALSE}
# ROOT DIRECTORY (to modify on your computer)
path.root <- "C:/Users/Gilar/OneDrive/Escritorio/Master/5_Semester/Thesis/MetaIBS"
path.mars  <- file.path(path.root, "Analysis/Scripts/Individual/Mars-2020")
path.data <- file.path(path.root, "data/phyloseq-objects")
path.plots <- file.path(path.root, "Analysis/Plots/Individual/Mars-2020")
```


## Read data
```{r read-data, echo=FALSE}
physeq <- readRDS(file.path(path.data, "physeq_mars.rds")) # phyloseq object


metadata <- physeq%>%ps_filter(Collection=="1st")
```


**************************
# 2. DESCRIPTIVE ANAYLISIS
**************************

## 2.1. Demographic and Clinical Information

```{r descriptive-analysis, echo=FALSE}

# Number of disease VS healthy (%)
metadata@sam_data%>%
  group_by(host_disease)%>%
  count()

# Age mean and sd per disease status
metadata@sam_data%>%
  group_by(host_disease)%>%
  summarize(mean = mean(host_age), 
            sd = sd(host_age))

# Histogramm of age per disease status
ggplot(metadata@sam_data, aes(host_age))+
  geom_histogram(aes(fill=host_disease))+
  facet_grid(host_disease~.)

# Number of male VS female
metadata@sam_data%>%
  group_by(host_sex)%>%
  count()

# Bar chart of sex categories per disease status
metadata@sam_data$host_sex <- factor(metadata@sam_data$host_sex,
                                  levels=c("male", "female"))

ggplot(metadata@sam_data, aes(x = host_sex)) + 
  geom_bar(aes(x = host_sex, fill = factor(host_sex))) +
  facet_grid(host_disease~.)+
  geom_text(aes(label=after_stat(count)),stat='count',position=position_dodge(0.9),vjust=-0.3)+
  xlab("Sex Category")+
  ylab("count")

# Histogramm of bmi per disease status
ggplot(metadata@sam_data, aes(host_bmi))+
  geom_density(aes(fill=host_disease))+
  expand_limits(x=0)+
  facet_grid(host_disease~.)

```


*****************************
# 3. ALPHA DIVERSITY ANALYSIS
*****************************

## 3.1. Rarefaction Curves

```{r rarefaction-curve, echo=FALSE}

# number of species against the number of samples
# This curve is created by randomly re-sampling the pool of N samples several times and then plotting the average number of species found on each sample

set.seed(1024)
#We work just with the 1st collection
rareres <- get_rarecurve(obj=physeq%>%ps_filter(Collection=="1st"), chunks=400) # chunks: number of subsample in a sample
# Rarefaction curve per disease status
prare1 <- ggrarecurve(obj=rareres, factorNames="host_disease",
                      indexNames=c("Observe", "Chao1", "ACE")
                      ) +
          scale_fill_manual(values=c("#00AED7", "#FD9347"))+
          scale_color_manual(values=c("#00AED7", "#FD9347"))+
          theme_bw()+
          theme(axis.text=element_text(size=8), panel.grid=element_blank(),
                strip.background = element_rect(colour=NA,fill="grey"),
                strip.text.x = element_text(face="bold"))          
prare1

# Rarefaction curve per sample per disease status
prare2 <- ggrarecurve(obj=rareres,
                      factorNames="host_disease",
                      shadow=FALSE,
                      indexNames=c("Observe", "Chao1", "ACE")
                      ) +
          scale_color_manual(values=c("#00AED7", "#FD9347"))+
          theme_bw()+
          theme(axis.text=element_text(size=8), panel.grid=element_blank(),
                strip.background = element_rect(colour=NA,fill="grey"),
                strip.text.x = element_text(face="bold"))
prare2

```


## 3.2. Analysis of Alpha Index

```{r alpha-diversity, echo=FALSE}

#   measure of microbiome diversity applicable to a single sample
#  for alpha diversity, many indices exist, each reflecting different aspects of community heterogeneity

alphaobj <- get_alphaindex(physeq%>%ps_filter(Collection=="1st"))

head(as.data.frame(alphaobj))

p_alpha <- ggbox(alphaobj, geom="violin", factorNames="host_disease") +
           scale_fill_manual(values=c("#00AED7", "#FD9347"))+
           theme(strip.background = element_rect(colour=NA, fill="grey"))
p_alpha

```

**********************************
# 4. TAXONOMY COMPOSITION ANALYSIS
**********************************

## Statistics and visualization of specific levels (Composition)

## 4.1. Class Analysis

```{r plot-classes}
class.taxa <- get_taxadf(obj=physeq, taxlevel=3)

# Relative Abundance of the 30 most abundant taxonomy will be visualized by default (parameter `topn=30`). 
rel.ab_class <- ggbartax(obj=class.taxa, facetNames="host_disease") +
          xlab(NULL) +
          ylab("relative abundance (%)") +
          guides(fill= guide_legend(keywidth = 0.5, keyheight = 0.5))

rel.ab_class


# Show the relative abundance in different groups.
rel.ab.group_class <- ggbartax(obj=class.taxa, facetNames="host_disease",
                               plotgroup=TRUE, topn=15) +
          xlab(NULL) +
          ylab("relative abundance (%)") +
          guides(fill= guide_legend(keywidth = 0.5, keyheight = 0.5, ncol=2))

rel.ab.group_class


#Absolute abundance
abs.ab_class <- ggbartax(obj=class.taxa, count=TRUE, facetNames="host_disease") +
          xlab(NULL) +
          ylab("count reads") +
          guides(fill= guide_legend(keywidth = 0.5, keyheight = 0.5))

abs.ab_class


# Show the absolute  in different groups.
abs.ab.group_class <- ggbartax(obj=class.taxa, count=TRUE, facetNames="host_disease",
                               plotgroup=TRUE,topn=15) +
          xlab(NULL) +
          ylab("count reads") +
          guides(fill= guide_legend(keywidth = 0.5, keyheight = 0.5))

abs.ab.group_class

```

## 4.2. Phylum Analysis

```{r plot-phylum, fig.width = 10, fig.height = 4}
phylum.taxa <- get_taxadf(obj=physeq, taxlevel=2)

rel.ab_phylum <- ggbartax(obj=phylum.taxa, facetNames="host_disease") +
          xlab(NULL) +
          ylab("relative abundance (%)") +
          guides(fill= guide_legend(keywidth = 0.5, keyheight = 0.5))

rel.ab_phylum


# Show the abundance in different groups.
rel.ab.group_phylum <- ggbartax(obj=phylum.taxa, facetNames="host_disease",
                               plotgroup=TRUE, topn=15) +
          xlab(NULL) +
          ylab("relative abundance (%)") +
          guides(fill= guide_legend(keywidth = 0.5, keyheight = 0.5, ncol=2))

rel.ab.group_phylum


#Absolute abundance
abs.ab_phylum <- ggbartax(obj=phylum.taxa, count=TRUE, facetNames="host_disease") +
          xlab(NULL) +
          ylab("count reads") +
          guides(fill= guide_legend(keywidth = 0.5, keyheight = 0.5))

abs.ab_phylum


# Show the absolute  in different groups.
abs.ab.group_phylum <- ggbartax(obj=phylum.taxa, count=TRUE,
                                facetNames="host_disease", plotgroup=TRUE,topn=15) +
          xlab(NULL) +
          ylab("count reads") +
          guides(fill= guide_legend(keywidth = 0.5, keyheight = 0.5))

abs.ab.group_phylum

```


## 4.3. Per Subject Analysis

```{r plot-id, fig.width = 10, fig.height = 4}

rel.id_phylum <- ggbartax(obj=phylum.taxa, facetNames="host_ID") +
          xlab("Run") +
          ylab("relative abundance (%)") +
          #facet_grid(rows = vars(host_disease))+
          guides(fill= guide_legend(keywidth = 0.5, keyheight = 0.5))

rel.id_phylum

#Absolute abundance
abs.id_phylum <- ggbartax(obj=phylum.taxa, count=TRUE, facetNames="host_ID") +
          xlab("Run") +
          ylab("count reads") +
          facet_grid(host_disease~.)+
          guides(fill= guide_legend(keywidth = 0.5, keyheight = 0.5))

abs.id_phylum

```

***********************************
# 5. FIRMICUTES/BACTEROIDETES RATIO
***********************************

```{r firmicutes/bacteroidetes ratio, echo=FALSE}

# First, relative abundances are calculated. Average per group for each phylum respectively
relevant.covariates <- c('Sample', 'Abundance', 'host_disease', 'Phylum', 'host_ID', 'host_subtype', 'Collection', 'host_sex')

# Agglomerate to phylum & genus levels
phylum.table <- physeq %>%
  tax_glom(taxrank = "Phylum") %>%                     # agglomerate at phylum level
  transform_sample_counts(function(x) {x/sum(x)} ) %>% # Transform to rel. abundance
  psmelt()                                             # Melt to long format

# Extract abundance of only Bacteroidota and Firmicutes
bacter <- phylum.table %>%
  filter(Phylum == "Bacteroidota") %>%
  select(all_of(relevant.covariates)) %>%
  dplyr::rename(Bacteroidota = Abundance) %>%
  select(-Phylum)

firmi <- phylum.table %>%
  filter(Phylum == "Firmicutes") %>%
  select(all_of(relevant.covariates)) %>%
  dplyr::rename(Firmicutes = Abundance) %>%
  select(-Phylum)

# Calculate log2 ratio Firmicutes/Bacteroidota
ratio.FB <- left_join(x=bacter, y=firmi, by=c('Sample', 'host_disease', 'host_ID', 'host_subtype', 'Collection', 'host_sex')) %>%
  relocate(Firmicutes, .after=Bacteroidota) %>%
  # Compute log ratios
  mutate(logRatioFB = log2(Firmicutes/Bacteroidota))

# Plot FB Ratio
ggplot(ratio.FB, aes(x = host_disease, y = logRatioFB))+
  geom_violin(aes(fill=host_disease))+
  scale_fill_manual(values=scales::alpha(c("blue", "red"), .3))+
  geom_jitter(width=0.1)+
  geom_signif(comparisons = list(c("Healthy", "IBS")), map_signif_level = TRUE, test="wilcox.test") +
  labs(x = "",  y = 'Log2(Firmicutes/Bacteroidota)', title = "F/B ratio")+
  theme_cowplot()+
  theme(legend.position="none")
# ggsave(file.path(path.fukui, "plot_ratioFB.jpg"), width=4, height=6)

# Statistical test
wilcox.test(ratio.FB[ratio.FB$host_disease == "IBS","logRatioFB"],
            ratio.FB[ratio.FB$host_disease == "Healthy","logRatioFB"]) # p = 0.6


#Looking at the F/B ratio in different time points
ggplot(ratio.FB, aes(x = host_disease, y = logRatioFB))+
  geom_violin(aes(fill=host_disease))+
  scale_fill_manual(values=scales::alpha(c("blue", "red"), .3))+
  geom_jitter(width=0.1)+
  facet_wrap(~Collection) +
  geom_signif(comparisons = list(c("Healthy", "IBS")), map_signif_level = TRUE, test="wilcox.test") +
  labs(x = "",  y = 'Log2(Firmicutes/Bacteroidota)')+
  theme_cowplot()+
  theme(legend.position="none")

```


******************
# 6. BETA ANALYSIS
******************

## 6.1. PCA Analysis

```{r pca-analysis, fig.width = 10, fig.height = 4}

# If the input was normalized, the method parameter should be settled NULL.
pcares <- get_pca(obj=physeq%>%ps_filter(Collection=="1st"), method="hellinger")

# Visualizing the result
pcaplot1 <- ggordpoint(obj=pcares, biplot=TRUE, speciesannot=TRUE,
                      factorNames=c("host_disease"), ellipse=TRUE) +
            scale_color_manual(values=c("#00AED7", "#FD9347")) +
            scale_fill_manual(values=c("#00AED7", "#FD9347"))

# pc = c(1, 3) to show the first and third principal components.
pcaplot2 <- ggordpoint(obj=pcares, pc=c(1, 3), biplot=TRUE, speciesannot=TRUE,
                       factorNames=c("host_disease"), ellipse=TRUE) +
            scale_color_manual(values=c("#00AED7", "#FD9347")) +
            scale_fill_manual(values=c("#00AED7", "#FD9347"))
pcaplot1 | pcaplot2

```


## 6.2. PCoA Analysis

```{r pcoa-analysis, fig.width = 10, fig.height = 4}

# distmethod
# "unifrac",  "wunifrac", "manhattan", "euclidean", "canberra", "bray", "kulczynski" ...(vegdist, dist)
pcoares <- get_pcoa(obj=physeq%>%ps_filter(Collection=="1st"), distmethod="bray", method="hellinger")

# Visualizing the result
pcoaplot1 <- ggordpoint(obj=pcoares, biplot=TRUE, speciesannot=TRUE,
                       factorNames=c("host_disease"), ellipse=TRUE) +
            scale_color_manual(values=c("#00AED7", "#FD9347")) +
            scale_fill_manual(values=c("#00AED7", "#FD9347"))

# first and third principal co-ordinates
pcoaplot2 <- ggordpoint(obj=pcoares, pc=c(1, 3), biplot=TRUE, speciesannot=TRUE,
                        factorNames=c("host_disease"), ellipse=TRUE) +
             scale_color_manual(values=c("#00AED7", "#FD9347")) +
             scale_fill_manual(values=c("#00AED7", "#FD9347"))

pcoaplot1 | pcoaplot2

```


## 6.3. Permuatational Multivariate Anaylsis of Variance 

```{r perm-multi-analysis, echo=FALSE}

distme <- get_dist(physeq%>%ps_filter(Collection=="1st"), distmethod ="bray", method="hellinger")
sampleda <- data.frame(sample_data(physeq%>%ps_filter(Collection=="1st")), check.names=FALSE)
sampleda <- sampleda[match(colnames(as.matrix(distme)),rownames(sampleda)),,drop=FALSE]
sampleda$host_disease <- factor(sampleda$host_disease)
set.seed(1024)
adores <- adonis(distme ~ host_disease, data=sampleda, permutation=9999)
data.frame(adores$aov.tab)

```


## 6.4. Hierarchical Cluster Analysis of Samples

```{r hierarchical-cluster, fig.width = 10, fig.height = 4}

hcsample <- get_clust(obj=physeq%>%ps_filter(Collection=="1st"), distmethod="bray",
                      method="hellinger", hclustmethod="average")
# rectangular layout
cplot1 <- ggclust(obj=hcsample,
                  layout = "rectangular",
                  pointsize=1,
                  fontsize=0,
                  factorNames=c("host_disease")
              ) +
              scale_color_manual(values=c("#00AED7", "#FD9347")) +
              theme_tree2(legend.position="right",
              plot.title = element_text(face="bold", lineheight=25,hjust=0.5))
# circular layout
cplot2 <- ggclust(obj=hcsample,
                  layout = "circular",
                  pointsize=1,
                  fontsize=2,
                  factorNames=c("host_disease")
          ) +
          scale_color_manual(values=c("#00AED7", "#FD9347")) +
          theme(legend.position="right")
cplot1 | cplot2

```

